lib/publiccloud/basetest.pm

```diff
@@ -51,16 +53,42 @@ sub provider_factory {
     else {
         die('Unknown PUBLIC_CLOUD_PROVIDER given');
     }
+
+    $provider->init();
+    $self->{provider} //= [];
+    push(@{$self->{provider}}, $provider);
+    return $provider;
 }
 
-sub get_image_id {
-    my ($self, $provider) = @_;
-    my $img_url    = get_required_var('PUBLIC_CLOUD_IMAGE_LOCATION');
-    my ($img_name) = $img_url =~ /([^\/]+)$/;
-    my $image_id   = $provider->find_img($img_name);
-    die("Image $img_name is not available in the cloud provider") unless ($image_id);
-    set_var('PUBLIC_CLOUD_IMAGE_ID', $image_id);
-    return $image_id;
+sub cleanup {
+    # to be overridden by tests
+    return;
 }
 
+sub _cleanup {
+    my ($self) = @_;
+    die("Cleanup called twice!") if ($self->{cleanup_called});
```

Comment:
if it has been called for the second time, shall we just return and not throw error with die()?

Reply:
It was meant as a assert(). Maybe return is enough, yes

Reply:
actually I like idea of dying here. We didn't expecting to call it more than once , so if that happened this mean we have bug in our logic and if you just silently return you will hide it

Reply:
agree, I will keep it.